### ISSUE TITLE
Open MPI 5/main bind all processes to single CPU (on multiple CPU system) when `-N` option is used. 

### ISSUE THREAD

**Issue by wzamazon** [open-mpi/ompi#11726]
Thank you for taking the time to submit an issue!

### What version of Open MPI are you using? (e.g., v3.0.5, v4.0.2, git branch name and hash, etc.)

both main and 5.0.x has this issue

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)

compile from source

### Please describe the system on which you are running

* Operating system/version: ubuntu 20
* Computer hardware: Intel CPU
* Network type: EFA

-----------------------------

## Details of the problem

My compute instance has 2 CPUs, but openmpi (prrte) will bind all processes to 1st CPU if I used `--map-by ppr:X:node` or `-N X`.

I used the following command to print bindings

```
ompi/main/mpirun  -n 16 -N 8  --report-bindings  --hostfile /path/to/hostfile    echo "Hello, world"
```
and got the following output:
```
[queue-p4d24xlarge-st-p4d24xlarge-2:183848] Rank 8 bound to package[0][core:0-23]
[queue-p4d24xlarge-st-p4d24xlarge-2:183848] Rank 9 bound to package[0][core:0-23]
[queue-p4d24xlarge-st-p4d24xlarge-2:183848] Rank 10 bound to package[0][core:0-23]
[queue-p4d24xlarge-st-p4d24xlarge-2:183848] Rank 11 bound to package[0][core:0-23]
[queue-p4d24xlarge-st-p4d24xlarge-2:183848] Rank 12 bound to package[0][core:0-23]
[queue-p4d24xlarge-st-p4d24xlarge-2:183848] Rank 13 bound to package[0][core:0-23]
[queue-p4d24xlarge-st-p4d24xlarge-2:183848] Rank 14 bound to package[0][core:0-23]
[queue-p4d24xlarge-st-p4d24xlarge-2:183848] Rank 15 bound to package[0][core:0-23]
[queue-p4d24xlarge-st-p4d24xlarge-1:184313] Rank 0 bound to package[0][core:0-23]
[queue-p4d24xlarge-st-p4d24xlarge-1:184313] Rank 2 bound to package[0][core:0-23]
[queue-p4d24xlarge-st-p4d24xlarge-1:184313] Rank 3 bound to package[0][core:0-23]
[queue-p4d24xlarge-st-p4d24xlarge-1:184313] Rank 1 bound to package[0][core:0-23]
[queue-p4d24xlarge-st-p4d24xlarge-1:184313] Rank 4 bound to package[0][core:0-23]
[queue-p4d24xlarge-st-p4d24xlarge-1:184313] Rank 5 bound to package[0][core:0-23]
[queue-p4d24xlarge-st-p4d24xlarge-1:184313] Rank 6 bound to package[0][core:0-23]
[queue-p4d24xlarge-st-p4d24xlarge-1:184313] Rank 7 bound to package[0][core:0-23]
Hello, world
....
```

**Comment by wzamazon** [comment]
Interestingly, this issue currently only impact hwloc 2. because there is a bug in prrte code that handle hwloc 1 and 2 differently.

**Comment by wzamazon** [comment]
Note that this hehavior is different from v4, which used orte and will bind processes to different CPUs

**Comment by wzamazon** [comment]
> Looks like https://github.com/openpmix/prrte/pull/1751 that fixed the issue for you, and it's already been backported and merged to https://github.com/openpmix/prrte/pull/1752 and https://github.com/openpmix/prrte/pull/1753.
> So the action to Open MPI is to bump prrte pointers.

No, the PR did not fix the issue of binding to same socket. It fixed the inconsistency between hwloc 1 and hwloc 2.

Previously, only hwloc 2 has this issue. hwloc 1 does not have this problem (by mistake)

After bumping prrte pointers, both hwloc 1 and hwloc 2 will have this problem

**Comment by rhc54** [comment]
> When user specified map by node, all processes on the same node will be bound to the 1st package

Just to be clear, that is not exactly correct. We tight-pack the procs, so since you have 24 cpus on the 1st package, the 1st 24 procs will be bound there. The next 24 procs will be placed on the 2nd package. Rationale: you want to tight-pack the procs to maximize performance since shared memory works best when both procs are on the same package.

**Comment by wzamazon** [comment]
WRF also mix use MPI and OpenMP, and will also be impacted

**Comment by rhc54** [comment]
FWIW: we have always told people who write multi-threaded code (whether using OpenMP or not) that they need to map-by package and bind-to package for their code to be performant. What you describe is someone not following that advice. Regardless of the default placement options used in OMPI, users doing what you describe are going to continually encounter performance problems.

Bottom line: they are getting bad advice on how to map/bind procs, whether using OMPI or MPICH - makes no difference.

**Comment by wzamazon** [comment]
> Regardless of the default placement options used in OMPI, users doing what you describe are going to continually encounter performance problems.

I am not sure that is the case.

As far as I know, Open MPI 5 is the outlier here.

Neither Open MPI 4 nor Intel MPI behavior the way Open MPI 5 does.

For example, the following example show Intel MPI's binding under such setting: 8 proceses per node, each node has 2 packages, each package has 24 cores:

```
(env) -bash-4.2$ mpirun -np 16 -ppn 8 --hostfile /fsx/PortaFiducia/hostfile /opt/intel/mpi/2021.6.0/bin/IMB-MPI1
[0] MPI startup(): Intel(R) MPI Library, Version 2021.6  Build 20220227 (id: 28877f3f32)
[0] MPI startup(): Copyright (C) 2003-2022 Intel Corporation.  All rights reserved.
[0] MPI startup(): library kind: release
[0] MPI startup(): libfabric version: 1.13.2rc1-impi
[0] MPI startup(): libfabric provider: efa
[0] MPI startup(): File "/opt/intel/mpi/2021.6.0/etc/tuning_skx_shm-ofi_efa_100.dat" not found
[0] MPI startup(): Load tuning file: "/opt/intel/mpi/2021.6.0/etc/tuning_skx_shm-ofi_efa.dat"
[0] MPI startup(): Rank    Pid      Node name                           Pin cpu
[0] MPI startup(): 0       93606    queue-p4d24xlarge-st-p4d24xlarge-1  {0,1,2,3,4,5}
[0] MPI startup(): 1       93607    queue-p4d24xlarge-st-p4d24xlarge-1  {6,7,8,9,10,11}
[0] MPI startup(): 2       93608    queue-p4d24xlarge-st-p4d24xlarge-1  {12,13,14,15,16,17}
[0] MPI startup(): 3       93609    queue-p4d24xlarge-st-p4d24xlarge-1  {18,19,20,21,22,23}
[0] MPI startup(): 4       93610    queue-p4d24xlarge-st-p4d24xlarge-1  {24,25,26,27,28,29}
[0] MPI startup(): 5       93611    queue-p4d24xlarge-st-p4d24xlarge-1  {30,31,32,33,34,35}
[0] MPI startup(): 6       93612    queue-p4d24xlarge-st-p4d24xlarge-1  {36,37,38,39,40,41}
[0] MPI startup(): 7       93613    queue-p4d24xlarge-st-p4d24xlarge-1  {42,43,44,45,46,47}
[0] MPI startup(): 8       67923    queue-p4d24xlarge-st-p4d24xlarge-2  {0,1,2,3,4,5}
[0] MPI startup(): 9       67924    queue-p4d24xlarge-st-p4d24xlarge-2  {6,7,8,9,10,11}
[0] MPI startup(): 10      67925    queue-p4d24xlarge-st-p4d24xlarge-2  {12,13,14,15,16,17}
[0] MPI startup(): 11      67926    queue-p4d24xlarge-st-p4d24xlarge-2  {18,19,20,21,22,23}
[0] MPI startup(): 12      67927    queue-p4d24xlarge-st-p4d24xlarge-2  {24,25,26,27,28,29}
[0] MPI startup(): 13      67928    queue-p4d24xlarge-st-p4d24xlarge-2  {30,31,32,33,34,35}
[0] MPI startup(): 14      67929    queue-p4d24xlarge-st-p4d24xlarge-2  {36,37,38,39,40,41}
[0] MPI startup(): 15      67930    queue-p4d24xlarge-st-p4d24xlarge-2  {42,43,44,45,46,47}
```

**Comment by rhc54** [comment]
I didn't say that the default placement options were the same across MPIs or even OMPI releases - they are not. Indeed, OMPI itself has varied considerably (sometimes reversing course) between its releases.

What I said was that anyone not mapping/binding by package - whether default or not - is going to have problems with multi-threaded applications. Your comments about WRF and GROMACS reflect bad advice in placement strategies if people aren't aware of that situation and blindly assume that the _default_ policy is going to give them best behavior.

Note that the default you show is in fact _not_ the correct pattern for non-multi-threaded apps! Which was the point you AWS folks made that caused us to change the OMPI behavior for v5!

Bottom line: you folks need to resolve the defaults in your own community and reflect them in YOUR schizo component. Don't blame PRRTE for your internal disagreements.

**Comment by wzamazon** [comment]
> Note that the default you show is in fact not the correct pattern for non-multi-threaded apps! Which was the point you AWS folks made that caused us to change the OMPI behavior for v5!

Sorry, I am not aware of such issue from AWS side. maybe you can give a pointer.

>Bottom line: you folks need to resolve the defaults in your own community and reflect them in YOUR schizo component. Don't blame PRRTE for your internal disagreements.

I do not think PRRTE is to be blamed. In fact, I do not think Open MPI should be blamed.

From my side, I observed a change of behavior between Open MPI 4 and Open MPI 5, which can cause the performance of Open MPI 5 to be 10x slower than that of Open MPI 4 under default setting, and wanted the OMPI community to be aware of it.

If the community agreed that this behavior change is better, please feel free to close this issue.

**Comment by rhc54** [comment]
> Sorry, I am not aware of such issue from AWS side. I would appreciate if you can give a pointer.

The issue was raised that OMPI v4 was providing default placements that were not optimal for single-threaded code. Such applications were seeing a significantly lower performance than expected. Root cause was that we weren't dense packing the procs so that they could efficiently use shared memory. Shared memory between two processes on different packages is much slower than if they are on the same package.

So the insistence was that we change the default placement strategy to optimize single-threaded performance. This is what some earlier versions of OMPI had done (as I said, OMPI has gone back and forth on this issue multiple times for years).

Negative impact of that was on multi-threaded codes, which really want to bind to a package - or better yet, to bind each proc to a number of cpu's equal to the number of threads it uses. I had in fact implemented just that (we detected the OpenMP directive and automatically bound the proc to that number of cpus) several years ago - and got my head bit off by the OMPI community. So we settled on providing the `map-by package bind-to package` options and told multi-thread app users to utilize them.

At least for OMPI v3, that was the ORTE defaults (I forget how far back that gyration went). Then folks decided that multi-threaded users were becoming more common, and that such users weren't realizing the need to properly define their placement policies. So....we changed to the v4 defaults.

And then we changed back again for v5.

As you can see, it has been a frustrating time as OMPI continues to bounce back and forth on this issue - both for OMPI's users and for me as the RTE developer. Which is why I created the OMPI schizo component so I could distance myself from further gyrations.

**Comment by rhc54** [comment]
Only caution is that changing it back to ORTE v4 behavior will be good for multi-threaded apps, but bad for single threaded apps - which is why it was changed in OMPI v5 (as previously noted). If you look back in the records, you'll find the single threaded folks complained bitterly about the changes in ORTE v4 as it messed them up. Look further back, and you'll see the multi-threaded folks complaining bitterly about the prior releases.

Can't please everyone, no matter which way you go. Best advice is to tell people _not_ to rely on default placement policies if performance is a concern - especially since nobody agrees on what is the best default (Slurm does it differently, as do various `mpirun` versions).

**Comment by darrylabbate** [comment]
I've only taken a cursory glance at this issue, but I wanted to bring attention to #11541, which I opened a few months back since it may be a good reference here

**Comment by lrbison** [comment]
Discussed today.  Plan will be to leave behavior as is, and ensure documentation is accurate.

**Comment by rhc54** [comment]
As mentioned on the PRRTE PR, we updated the default behaviors a bit while fixing the no-np behavior. The default map/bind logic is now set to bind by default to the object we were told to map by if no binding policy was specified, and we default to map by the bind object if binding was specified but no mapping. Defaults if neither are specified remain unchanged.
